File: src/shellgpt/assistant.py
"""OpenAI interaction with secure API key handling."""
from openai import OpenAI
from rich.console import Console
from rich.prompt import Prompt
from typing import Tuple
import sys

from .config import Config
from .shell import ShellContext

console = Console()

class Assistant:
    def __init__(self):
        self.config = Config()
        self.shell_context = ShellContext()
        
        try:
            api_key = self.config.get_api_key()
            self.client = OpenAI(api_key=api_key)
        except ValueError as e:
            console.print(f"\n[red]{str(e)}[/red]")
            sys.exit(1)

    def get_command_suggestion(self, query: str) -> Tuple[str, str]:
        """Get command suggestion from OpenAI."""
        recent_commands = self.shell_context.get_recent_commands()
        
        context = ""
        if recent_commands:
            context += "Recent commands:\n" + "\n".join(f"- {cmd}" for cmd in recent_commands)

        messages = [
            {
                "role": "system", 
                "content": """You are a command line assistant for zsh on MacOS. 
                Always respond in this format: EXPLANATION|COMMAND
                where EXPLANATION is one sentence explaining what the command does,
                and COMMAND is the exact command to run.
                Example: 'Shows last 10 commands from history|history 10'"""
            },
            {
                "role": "user", 
                "content": f"Context:\n{context}\n\nQuery: {query}"
            }
        ]

        try:
            response = self.client.chat.completions.create(
                model="gpt-4o-mini",
                messages=messages,
                temperature=0.3,
                max_tokens=150
            )
            
            content = response.choices[0].message.content
            if '|' in content:
                explanation, command = content.split('|', 1)
                return explanation.strip(), command.strip()
            return content.strip(), ""
            
        except Exception as e:
            console.print(f"[red]Error: {str(e)}[/red]")
            if 'authentication' in str(e).lower() or 'api key' in str(e).lower():
                console.print("\n[yellow]This appears to be an API key issue. Please check that your OPENAI_API_KEY environment variable is set correctly.[/yellow]")
            return "Error occurred", ""

    def interactive_chat(self):
        """Start an interactive chat session."""
        console.print("[blue]Starting interactive chat session (type 'exit' to quit)[/blue]")
        
        messages = [
            {"role": "system", "content": "You are a helpful command line assistant for MacOS. You can help with shell commands, explain terminal concepts, and provide general assistance."}
        ]
        
        while True:
            query = Prompt.ask("\nYou")
            
            if query.lower() in ('exit', 'quit'):
                break
                
            try:
                messages.append({"role": "user", "content": query})
                
                response = self.client.chat.completions.create(
                    model="gpt-4o-mini",
                    messages=messages,
                    temperature=0.7
                )
                
                assistant_message = response.choices[0].message.content
                console.print(f"\n[green]Assistant:[/green] {assistant_message}")
                
                messages.append({"role": "assistant", "content": assistant_message})
                
            except Exception as e:
                console.print(f"\n[red]Error: {str(e)}[/red]")

File: src/assistant.rs
use anyhow::Result;
use async_openai::types::{
    ChatCompletionRequestAssistantMessageArgs, ChatCompletionRequestSystemMessageArgs,
    ChatCompletionRequestUserMessageArgs, CreateChatCompletionRequestArgs,
};
use async_openai::{config::OpenAIConfig, Client};
use console::Style;
use dialoguer::Input;
use indicatif::{ProgressBar, ProgressStyle};

use crate::config::get_api_key;
use crate::shell::ShellContext;

pub static OUTPUT_DELIMITER: &str = "§";

pub fn create_client(shell_context: &ShellContext) -> Result<Client<OpenAIConfig>> {
    let api_key = get_api_key(&shell_context.shell_type)?;
    Ok(Client::with_config(OpenAIConfig::new().with_api_key(api_key)))
}

pub async fn get_command_suggestion(
    client: &Client<OpenAIConfig>,
    shell_context: &ShellContext,
    query: &str,
    context: Option<&str>,
) -> Result<(String, String)> {
    let mut context_str = String::new();

    // Add piped input context if provided
    if let Some(piped_context) = context {
        context_str.push_str("Input context:\n");
        context_str.push_str(piped_context);
        context_str.push_str("\n\n");
    };

    let system_prompt = get_command_suggestion_prompt(shell_context);

    let request = CreateChatCompletionRequestArgs::default()
        .model("gpt-4o-mini")
        .messages([
            ChatCompletionRequestSystemMessageArgs::default()
                .content(system_prompt)
                .build()?
                .into(),
            ChatCompletionRequestUserMessageArgs::default()
                .content(format!("Context:\n{:#?}\n\nQuery: {}", context, query))
                .build()?
                .into(),
        ])
        .temperature(0.3)
        .max_tokens(150u16)
        .build()?;

    let response = client.chat().create(request).await?;

    if let Some(choice) = response.choices.first() {
        if let Some(content) = &choice.message.content {
            if let Some((explanation, command)) = content.split_once(OUTPUT_DELIMITER) {
                return Ok((explanation.trim().to_string(), command.trim().to_string()));
            }
            return Ok((content.trim().to_string(), String::new()));
        }
    }

    Ok((String::new(), String::new()))
}

pub async fn interactive_chat(
    client: &Client<OpenAIConfig>,
    shell_context: &ShellContext,
) -> Result<()> {
    println!(
        "{}",
        Style::new()
            .blue()
            .apply_to("Starting interactive chat session (type 'exit' to quit)")
    );

    let system_prompt = format!(
        "You are a helpful command line assistant for {}. \
                You can help with shell commands, explain terminal concepts, \
                and provide general assistance.",
        shell_context.os
    );

    let mut messages = vec![ChatCompletionRequestSystemMessageArgs::default()
        .content(system_prompt)
        .build()?
        .into()];

    let spinner = ProgressBar::new_spinner();
    spinner.set_style(
        ProgressStyle::default_spinner()
            .tick_chars("⠁⠂⠄⡀⢀⠠⠐⠈")
            .template("{spinner} Thinking...")?,
    );

    loop {
        let query: String = Input::new().with_prompt("You").interact()?;

        if query.eq_ignore_ascii_case("exit") || query.eq_ignore_ascii_case("quit") {
            break;
        }

        messages.push(
            ChatCompletionRequestUserMessageArgs::default()
                .content(query)
                .build()?
                .into(),
        );

        spinner.enable_steady_tick(std::time::Duration::from_millis(80));

        let request = CreateChatCompletionRequestArgs::default()
            .model("gpt-4o-mini")
            .messages(messages.clone())
            .temperature(0.7)
            .build()?;

        let response = client.chat().create(request).await?;

        spinner.disable_steady_tick();

        if let Some(choice) = response.choices.first() {
            if let Some(content) = &choice.message.content {
                println!(
                    "\n{} {}",
                    Style::new().green().apply_to("Assistant:"),
                    content
                );

                messages.push(
                    ChatCompletionRequestAssistantMessageArgs::default()
                        .content(content.clone())
                        .build()?
                        .into(),
                );
            }
        }
    }

    Ok(())
}

fn get_command_suggestion_prompt(shell_context: &ShellContext) -> String {
    let role_and_env_assertion = format!(
        "You are a command line assistant for {} on {}.",
        shell_context.shell_type, shell_context.os
    );

    let response_format = format!(
        "Always respond in this format: EXPLANATION{OUTPUT_DELIMITER}COMMAND"
    );

    let format_description = "
        where EXPLANATION is one sentence explaining what the command does, \
        and COMMAND is the exact command to run.
        ";

    let prompt_example = format!(
        "Example: 'Shows last 10 commands from history{OUTPUT_DELIMITER}history 10'"
    );

    format!(
        "{}\n{}\n{}\n{}",
        role_and_env_assertion, response_format, format_description, prompt_example
    )
}
